### ISSUE TITLE
Why does the documentation not include the Exceptions that may be thrown??

### ISSUE THREAD

**Issue by bdcoder2** [JamesNK/Newtonsoft.Json#1685]
The NetwtonSoft documentation seems to lack a LOT of detail.  For example:

The documentation for **JsonConvert.DeserializeObject<T> Method (String)**  (which I picked at random), found at:
https://www.newtonsoft.com/json/help/html/M_Newtonsoft_Json_JsonConvert_DeserializeObject__1.htm

The documentation makes NO mention of the exceptions that may be thrown by this method, if any.  However, if an invalid JSON string is passed to this method, the following exception is thrown:

`Newtonsoft.Json.JsonReaderException`

So, in reality this method should be wrapped in a try/catch when used.

I'll apologize in advance if I missed some preamble in the documentation that indicated all methods can throw an exception, but I find more and more "documentation" these days contains less and less details and I find myself wrapping all third party code in try/catch -- just because the documentation makes no mention of what exceptions (if any) may be raised.

**Comment by bdcoder2** [comment]
How disappointing (but not surprising) in this day and age that great software lacks the most important thing -- great documentation.

ps:  Because I don't have a document of acronyms, I have no idea what a "PR" is 😁

**Comment by cyberphone** [comment]
@bdcoder2 "PR" means Pull Request which is a Git way of introducing potential updates.

Regarding thrown exceptions, I prefer catching all exceptions in a few well selected places but I guess it depends on what kind of application you are working with...

**Comment by bdcoder2** [comment]
cyberphone, Just curious, how do you know where to catch exceptions if the documentation for whatever third party code you are using does NOT tell you if an exception can be thrown?

**Comment by cyberphone** [comment]
Hi @bdcoder2. I may be lazy but I just don't do any assumptions because third party code usually depends on OS/Platform features which in turn may cast exceptions as well.  If there is a specific error condition you need to deal with in a unique way things become considerably more complex.

_Where_ you catch exceptions is pretty application dependent, in my Web server stuff I surround the entire request code with a single try cast and return an HTTP 400 (Bad Request) + short message if something went wrong.

In Java there are "checked" (declared) exceptions but I find them mainly to be a problem for me as a designer of third party code.

**Comment by bdcoder2** [comment]
Agreed - I was simply trying t point out that the lack of documentation with regard to exceptions makes it IMPOSSIBLE to code efficiently.  It has gotten to the point that I have to wrap third party calls in try/catch; even when it may not be necessary -- because I have NO documentation to tell me otherwise; but must code defensively.

The bottom line is, if you create libraries that are meant to be used by others, provide clear and detailed documentation.   After 40 years of programming, I have witnessed the downfall of useful documentation.

**Comment by cyberphone** [comment]
As a true old-timer I agree that documentation has become less useful but it also has to do with the _size_, _dependencies_ and _fluidity_ of modern software.  OTOH,  Googling + Open Source + Stackoverflow + GitHub feels like a much more productive documentation system than was available for the Win32 API in the mid 90'ties.

Although clearly not my business, I don't fully get why you need to wrap calls to third-party software in try/catch.  Personally, I have mostly given up that idea and rather treat all code the same.  This is really the power of exception schemes over the previous (quite _error-prone_), C-style return code testing.

**Comment by bdcoder2** [comment]
"I don't fully get why you need to wrap calls to third-party software in try/catch." -- Because I have no idea if the third party code will cause my application to die and I want some control over flow and what happens.

Again, as we have both been around for such a long time, I am not so sure that having to Google + review open source code + asking questions on Stackoverflow is more efficient than being able to read documentation that clearly states "the following exceptions may be raised" or "No exceptions are raised".  

Life was simpler back then :) - Cheers.

**Comment by cyberphone** [comment]
With heavy layered and complex software, traditional error handling feels pretty unworkable.  Assume you have a `readDocument` command I would without hesitation do something like the following:
```c#
    try
    {
        doc = readDocument();
    }
    catch (Exception ex)
    {
        // Messagebox/alert the user etc, gracefully exit/cancel/throw if needed
        // "Document couldn't be read because: xxxxxxxxxx"
        // where xxxxxxxxx comes from the Exception
    }
```
If there is some very common error you can add a specific exception handler for that as well,  However, using the scheme above _this is just an option_.  Yep, a bit "raw" but the alternative, trying to figure all the error conditions worth dealing with separately comes at a price.  I won't pay 😂

**Comment by bdcoder2** [comment]
Agreed  -- that is what I am currently doing with all third-party software -- wrapping in try/catch -- but my original question is WHY SHOULD I HAVE TO?   All I am asking is that developers take the extra 20 seconds to include a sentence or note that says (for whatever function they are "documenting":

`readDocument() does NOT throw any exceptions.`

I guess one could take the ultimate approach and have every program look like this -- simply devolve to a single try/catch statement and hope for the best :)

```
main()
{
   try
   {
      do_everything();
   }
   catch ( Exception ex)
   {
      // Messagebox/alert( "A program error occurred (no clue as to why) -- too bad for you.");
   }
}
```
